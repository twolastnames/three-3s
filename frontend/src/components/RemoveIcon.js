import React from 'react';
import PropTypes from 'prop-types';

import { fetchWithMessages } from '../helpers/fetchWithMessages';

const patch = fetchWithMessages({ method: 'PATCH' });

const d =
  'M15 4H5v16h14V8h-4V4zM3 2.992C3 2.444 3.447 2 3.999 2H16l5 ' +
  '5v13.993A1 1 0 0 1 20.007 22H3.993A1 1 0 0 1 3 21.008V2.992zM16 ' +
  '11v2H8v-2h8z';

const RemoveSvg = ({ onClick }) => (
  <svg
    onClick={onClick}
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 24 24"
    width="24"
    height="24"
  >
    <path fill="none" d="M0 0h24v24H0z" />
    <path d={d} />
  </svg>
);

RemoveSvg.propTypes = {
  onClick: PropTypes.func.isRequired,
};

const doRemove = (url, description, triggerUpdate) => async () => {
  await patch(description, url);
  triggerUpdate();
};

export function RemoveIcon({ url, description, triggerUpdate = () => {} }) {
  return RemoveSvg({ onClick: doRemove(url, description, triggerUpdate) });
}

RemoveIcon.propTypes = {
  url: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  triggerUpdate: PropTypes.func,
};
