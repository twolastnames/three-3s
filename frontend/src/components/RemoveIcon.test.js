import React from 'react';

import { shallow } from 'enzyme';
import { RemoveIcon } from './RemoveIcon';

describe('RemoveIcon', () => {
  it('renders visually', () => {
    const wrapper = shallow(
      <RemoveIcon
        url="/threAS3/suites/5?remove_scenario_id=7"
        description="a test"
        acknowledgeDelete={() => {}}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('a successful response', () => {
    let triggerUpdate = jest.fn();

    beforeAll(() => {
      withFetch().mockOk({});
      const wrapper = shallow(
        <RemoveIcon
          url="/my/url"
          description="a test"
          triggerUpdate={triggerUpdate}
        />
      );
      wrapper.find('svg').simulate('click');
    });

    it('gives an info acknowledgement', () => {
      expect(triggerUpdate).toBeCalledTimes(1);
      expect(triggerUpdate).toBeCalledWith();
    });

    it('is called with requested url', () => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith('/my/url', { method: 'PATCH' });
    });
  });

  describe('an unsuccessful response', () => {
    let triggerUpdate = jest.fn();

    beforeAll(() => {
      withFetch().mockNotOk(666);
      const wrapper = shallow(
        <RemoveIcon
          url="/my/url"
          description="a test"
          triggerUpdate={triggerUpdate}
        />
      );
      wrapper.find('svg').simulate('click');
    });

    it('gives an error acknowledgement', () => {
      expect(triggerUpdate).toBeCalledTimes(1);
      expect(triggerUpdate).toBeCalledWith();
    });

    it('is called with requested url', () => {
      expect(global.fetch).toBeCalledTimes(1);
      expect(global.fetch).toBeCalledWith('/my/url', { method: 'PATCH' });
    });
  });

  describe('an exception', () => {
    let triggerUpdate = jest.fn();

    beforeAll(() => {
      withFetch().mockException('a message');
      const wrapper = shallow(
        <RemoveIcon
          url="/my/url"
          description="a test"
          triggerUpdate={triggerUpdate}
        />
      );
      wrapper.find('svg').simulate('click');
    });

    it('gives an error acknowledgement', () => {
      expect(triggerUpdate).toBeCalledTimes(1);
      expect(triggerUpdate).toBeCalledWith();
    });

    it('is called with requested url', () => {
      expect(global.fetch).toBeCalledWith('/my/url', { method: 'PATCH' });
      expect(global.fetch).toBeCalledTimes(1);
    });
  });
});
