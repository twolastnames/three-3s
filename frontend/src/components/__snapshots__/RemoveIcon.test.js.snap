// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveIcon renders visually 1`] = `ShallowWrapper {}`;
